client-misk/src/main/kotlin/app/cash/backfila/client/misk/internal/EmbeddedBackfillRun.kt

```diff
@@ -0,0 +1,187 @@
+        .dry_run(dryRun)
+        .build())
+    precomputeProgress = prepareBackfillResponse.instances.associate {
+      it.instance_name to MutableInstanceCursor(
+          it.instance_name, it.backfill_range)
+    }
+    scanProgress = prepareBackfillResponse.instances.associate {
+      it.instance_name to MutableInstanceCursor(
+          it.instance_name, it.backfill_range)
+    }
+  }
+
+  override fun precomputeScan(): GetNextBatchRangeResponse {
+    val notDone = precomputeProgress.filterNot { it.value.done }
+    check(notDone.isNotEmpty()) { "Nothing left to precompute" }
+    val cursor = notDone.values.first()
+    val response =
+        operator.getNextBatchRange(GetNextBatchRangeRequest.Builder()
+            .instance_name(cursor.instanceName)
+            .backfill_range(cursor.keyRange)
+            .previous_end_key(cursor.previousEndKey)
+            .parameters(parameters)
+            .batch_size(batchSize)
+            .scan_size(scanSize)
+            .compute_count_limit(computeCountLimit)
+            .precomputing(true)
+            .dry_run(dryRun)
+            .build())
+    if (response.batches.isEmpty()) {
+      cursor.done = true
+    } else {
+      cursor.previousEndKey = response.batches.map { it.batch_range.end }.max()
+      response.batches.onEach {
+        precomputeMatchingCount += it.matching_record_count
+        precomputeScannedCount += it.scanned_record_count
+      }
+    }
+    return response
+  }
+
+  override fun precomputeRemaining() {
+    do {
+      precomputeScan()
+    } while (!finishedPrecomputing())
+  }
+
+  override fun finishedPrecomputing() = precomputeProgress.all { it.value.done }
+
+  override fun instanceScan(instanceName: String): GetNextBatchRangeResponse {
+    val cursor = scanProgress.get(instanceName)!!
+    val response =
+        operator.getNextBatchRange(GetNextBatchRangeRequest.Builder()
+            .instance_name(cursor.instanceName)
+            .backfill_range(cursor.keyRange)
+            .previous_end_key(cursor.previousEndKey)
+            .parameters(parameters)
+            .batch_size(batchSize)
+            .scan_size(scanSize)
+            .compute_count_limit(computeCountLimit)
+            .dry_run(dryRun)
+            .build())
+    when (response.batches.isEmpty()) {
+      true -> cursor.done = true
+      false -> cursor.previousEndKey = response.batches.map { it.batch_range.end }.max()
+    }
+    batchesToRun.addAll(response.batches
+        .filterNot { it.matching_record_count == 0L } // Remove batches that have no matching records.
+        .map {
+          BatchSnapshot(instanceName,
+              it.batch_range, it.scanned_record_count,
+              it.matching_record_count)
+        })
+    return response
+  }
+
+  override fun singleScan(): GetNextBatchRangeResponse {
+    val cursor = scanProgress.values.first { !it.done }
+    return instanceScan(cursor.instanceName)
+  }
+
+  override fun scanRemaining() {
+    do singleScan()
```

Comment:
lack of {} on a do while looks really weird

Reply:
Ok. has brackets now.